# Task 1 — Load & Label the NSL-KDD Data

**Goal:** turn the raw, header-less data file into a labeled pandas DataFrame we can reason about.

**Why it matters:** raw, the file is just anonymous columns of numbers — a model can't learn from `column_4`, and neither can you. Naming the columns and defining a *target* (what we're predicting) is step one of every ML project. In security, it's also where we decide the question: *“is this network connection an attack, yes or no?”*

Fill in each `# TODO` cell, run it (Shift+Enter), and check the output. Try before peeking at `reference/data_loader.py`.

## Step 0 — Imports
You'll need the **pandas** library for working with tabular data. By convention it's imported with the alias `pd`.

In [1]:
# TODO: import pandas, aliased as pd
import pandas as pd

## Step 1 — The column names (provided — just run this cell)
The 41 feature names, in order, then the two trailing columns. Typing these by hand is error-prone, so they're given to you. Run the cell so the variables exist.

In [2]:
feature_columns = [
    "duration", "protocol_type", "service", "flag", "src_bytes", "dst_bytes",
    "land", "wrong_fragment", "urgent", "hot", "num_failed_logins", "logged_in",
    "num_compromised", "root_shell", "su_attempted", "num_root",
    "num_file_creations", "num_shells", "num_access_files", "num_outbound_cmds",
    "is_host_login", "is_guest_login", "count", "srv_count", "serror_rate",
    "srv_serror_rate", "rerror_rate", "srv_rerror_rate", "same_srv_rate",
    "diff_srv_rate", "srv_diff_host_rate", "dst_host_count", "dst_host_srv_count",
    "dst_host_same_srv_rate", "dst_host_diff_srv_rate",
    "dst_host_same_src_port_rate", "dst_host_srv_diff_host_rate",
    "dst_host_serror_rate", "dst_host_srv_serror_rate", "dst_host_rerror_rate",
    "dst_host_srv_rerror_rate",
]

# The full column list = 41 features + the two trailing columns.
all_columns = feature_columns + ["label", "difficulty"]
print(f"{len(all_columns)} column names ready")  # should print 43


43 column names ready


## Step 2 — Read the file
Read `../data/raw/KDDTrain+.txt` into a DataFrame called `df`.

**Hints:**
- Use `pd.read_csv(...)`.
- The file has **no header row** — so pass `header=None`, otherwise pandas would steal the first data row and use it as column titles.
- Give it your column names with the `names=` argument.
- The path is `../data/raw/KDDTrain+.txt` because this notebook lives inside the `notebooks/` folder (`..` means “go up one level”).

In [7]:
# TODO: df = pd.read_csv( path , header=..., names=... )
df=pd.read_csv("/Users/rohanb/06_projects/CYBERSECURITY/data/raw/KDDTrain+.txt", header=None, names=all_columns)
df.shape


(125973, 43)

## Step 3 — Look at it
Always eyeball your data after loading. Show the first few rows and the shape.

**Hints:** `df.head()` shows the top rows; `df.shape` gives `(rows, columns)`.

In [10]:
# TODO: display df.head() and print df.shape
df.head()


,duration,protocol_type,service,flag,src_bytes,dst_bytes,land,wrong_fragment,urgent,hot,...,dst_host_same_srv_rate,dst_host_diff_srv_rate,dst_host_same_src_port_rate,dst_host_srv_diff_host_rate,dst_host_serror_rate,dst_host_srv_serror_rate,dst_host_rerror_rate,dst_host_srv_rerror_rate,label,difficulty
0,0,tcp,ftp_data,SF,491,0,0,0,0,0,...,0.17,0.03,0.17,0.00,0.00,0.00,0.05,0.00,normal,20
1,0,udp,other,SF,146,0,0,0,0,0,...,0.00,0.60,0.88,0.00,0.00,0.00,0.00,0.00,normal,15
2,0,tcp,private,S0,0,0,0,0,0,0,...,0.10,0.05,0.00,0.00,1.00,1.00,0.00,0.00,neptune,19
3,0,tcp,http,SF,232,8153,0,0,0,0,...,1.00,0.00,0.03,0.04,0.03,0.01,0.00,0.01,normal,21
4,0,tcp,http,SF,199,420,0,0,0,0,...,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,normal,21


In [11]:
print(df.shape)

(125973, 43)


## Step 4 — Create the target column `is_attack`
Add a new column that is `0` when the connection is benign and `1` when it's an attack.

**Hints:**
- The `label` column contains the string `"normal"` for benign traffic, and various attack names otherwise.
- The comparison `df["label"] != "normal"` produces a column of `True`/`False`.
- Chaining `.astype(int)` turns `True`→`1` and `False`→`0`.
- Assign the result to `df["is_attack"]`.

In [17]:
# TODO: df["is_attack"] = ...
df["is_attack"] = (df["label"]!="normal").astype(int)

In [18]:
df

,duration,protocol_type,service,flag,src_bytes,dst_bytes,land,wrong_fragment,urgent,hot,...,dst_host_diff_srv_rate,dst_host_same_src_port_rate,dst_host_srv_diff_host_rate,dst_host_serror_rate,dst_host_srv_serror_rate,dst_host_rerror_rate,dst_host_srv_rerror_rate,label,difficulty,is_attack
0,0,tcp,ftp_data,SF,491,0,0,0,0,0,...,0.03,0.17,0.00,0.00,0.00,0.05,0.00,normal,20,0
1,0,udp,other,SF,146,0,0,0,0,0,...,0.60,0.88,0.00,0.00,0.00,0.00,0.00,normal,15,0
2,0,tcp,private,S0,0,0,0,0,0,0,...,0.05,0.00,0.00,1.00,1.00,0.00,0.00,neptune,19,1
3,0,tcp,http,SF,232,8153,0,0,0,0,...,0.00,0.03,0.04,0.03,0.01,0.00,0.01,normal,21,0
4,0,tcp,http,SF,199,420,0,0,0,0,...,0.00,0.00,0.00,0.00,0.00,0.00,0.00,normal,21,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
125968,0,tcp,private,S0,0,0,0,0,0,0,...,0.06,0.00,0.00,1.00,1.00,0.00,0.00,neptune,20,1
125969,8,udp,private,SF,105,145,0,0,0,0,...,0.01,0.01,0.00,0.00,0.00,0.00,0.00,normal,21,0
125970,0,tcp,smtp,SF,2231,384,0,0,0,0,...,0.06,0.00,0.00,0.72,0.00,0.01,0.00,normal,18,0
125971,0,tcp,klogin,S0,0,0,0,0,0,0,...,0.05,0.00,0.00,1.00,1.00,0.00,0.00,neptune,20,1


## Step 5 — Acceptance checks
Prove it worked.

**Hints:** print `df.shape`, and use `df["is_attack"].value_counts()` to count benign vs attack.

In [15]:
# TODO: print df.shape and df["is_attack"].value_counts()
print(df.shape)

(125973, 44)


In [16]:
df["is_attack"].value_counts()

is_attack
False    67343
True     58630
Name: count, dtype: int64

### ✅ You pass Task 1 when you see:
- Shape = **(125973, 44)**  — 41 features + label + difficulty + your new is_attack
- `is_attack` counts ≈ **67343 benign (0)** and **58630 attack (1)**

When you get there (or hit an error), paste the output to me for review.